frame/treasury/src/lib.rs

```diff
@@ -124,10 +295,183 @@ decl_module! {
 		/// Percentage of spare funds (if any) that are burnt per spend period.
 		const Burn: Permill = T::Burn::get();
 
+		/// The period for which a tip remains open after is has achieved threshold tippers.
+		const TipCountdown: T::BlockNumber = T::TipCountdown::get();
+
+		/// The amount of the final tip which goes to the original reporter of the tip.
+		const TipFindersFee: Percent = T::TipFindersFee::get();
+
+		/// The amount held on deposit for placing a tip report.
+		const TipReportDepositBase: BalanceOf<T> = T::TipReportDepositBase::get();
+
+		/// The amount held on deposit per byte within the tip report reason.
+		const TipReportDepositPerByte: BalanceOf<T> = T::TipReportDepositPerByte::get();
+
 		type Error = Error<T>;
 
 		fn deposit_event() = default;
 
+		/// Report something `reason` that deserves a tip and claim any eventual the finder's fee.
+		///
+		/// The dispatch origin for this call must be _Signed_.
+		///
+		/// Payment: `TipReportDepositBase` will be reserved from the origin account, as well as
+		/// `TipReportDepositPerByte` for each byte in `reason`.
+		///
+		/// - `reason`: The reason for, or the thing that deserves, the tip; generally this will be
+		///   a UTF-8-encoded URL.
+		/// - `who`: The account which should be credited for the tip.
+		///
+		/// Emits `NewTip` if successful.
+		///
+		/// # <weight>
+		/// - `O(R)` where `R` length of `reason`.
+		/// - One balance operation.
+		/// - One storage mutation (codec `O(R)`).
+		/// - One event.
+		/// # </weight>
+		fn report_awesome(origin, reason: Vec<u8>, who: T::AccountId) {
+			let finder = ensure_signed(origin)?;
+
+			const MAX_SENSIBLE_REASON_LENGTH: usize = 16384;
+			ensure!(reason.len() <= MAX_SENSIBLE_REASON_LENGTH, Error::<T>::ReasonTooBig);
+
+			let reason_hash = T::Hashing::hash(&reason[..]);
+			ensure!(!Reasons::<T>::exists(&reason_hash), Error::<T>::AlreadyKnown);
+			let hash = T::Hashing::hash_of(&(&reason_hash, &who));
+			ensure!(!Tips::<T>::exists(&hash), Error::<T>::AlreadyKnown);
+
+			let deposit = T::TipReportDepositBase::get()
+				+ T::TipReportDepositPerByte::get() * (reason.len() as u32).into();
+			T::Currency::reserve(&finder, deposit)?;
+
+			Reasons::<T>::insert(&reason_hash, &reason);
```

Comment:
It seems that in general you have relaxed your position with on-chain strings.

Just to clarify, so as long as a user places a deposit for the data they place on chain, it seems okay to put strings/arbitrary bytes on the blockchain? Following a rule like this in the general, do we run into an issues similar to the data bloat which is found with smart contract platforms (and is addressed with storage rent)?

I guess I wonder if this kind of code is more of a "solution for now", and would be changed in the future once we have better integration with services like IPFS.

The alternative here is off-chain coordination of the `reason` string, which seems much harder for a UI to manage.

Reply:
true; but for all instances where users can place data on-chain, there needs to be economic cost for each byte-block. this can be done through a per-time charge (rent) or through a deposit (where there's still a payment but through opportunity cost/inflation). deposits will be larger - probably be about 10x.

i don't think it's to be encouraged *in general*, but right now we have no option since there is no compelling and readily available alternative. it's not something i would rush into with polkadot mainnet but for kusama it seems ok.